Cython/Build/Tests/TestRecythonize.py

```diff
@@ -1,212 +1,218 @@
-import shutil
 import os
+import shutil
 import tempfile
 import time
 
 import Cython.Build.Dependencies
 import Cython.Utils
 from Cython.TestUtils import CythonTest
 
+SAME = "The result of cytonization is the same"
+INCORRECT = "Incorrect cythonization"
+LINE_BEFORE_IMPLEMENTATION = '  /* "{filename}":{at_line}\n'
+VARS_LINE = '  /*--- Wrapped vars code ---*/\n'
 
-def fresh_cythonize(*args, **kwargs):
-    Cython.Utils.clear_function_caches()
-    Cython.Build.Dependencies._dep_tree = None  # discard method caches
-    Cython.Build.Dependencies.cythonize(*args, **kwargs)
 
 class TestRecythonize(CythonTest):
+    language_level = 3
+    dep_tree = Cython.Build.Dependencies.create_dependency_tree()
+
+    def clear_function_and_Dependencies_caches(self):
+        Cython.Utils.clear_function_caches()
+        Cython.Build.Dependencies._dep_tree = None  # discard method caches
 
     def setUp(self):
         CythonTest.setUp(self)
+        self.clear_function_and_Dependencies_caches()
         self.temp_dir = (
             tempfile.mkdtemp(
                 prefix='recythonize-test',
                 dir='TEST_TMP' if os.path.isdir('TEST_TMP') else None
             )
         )
+        self.src_dir = os.path.join(self.temp_dir, 'src')
+        os.mkdir(self.src_dir)  # self.temp_dir is already a temp
 
     def tearDown(self):
         CythonTest.tearDown(self)
+        self.clear_function_and_Dependencies_caches()
         shutil.rmtree(self.temp_dir)
 
-    def test_recythonize_pyx_on_pxd_change(self):
+    def write_to_file(self, path, text):
+        try:
+            res = os.path.getmtime(path)
+        except OSError:  # not FileNotFoundError for compatibility
+            res = .0
+
+        with open(path, "w") as f:
+            f.write(text)
+
+        # Make sure the file has a newer timestamp,
+        # otherwise cythonize does not work as expected
+        # on Linux-like systems #4245
+        while 1:
+            if os.path.getmtime(path) != res:
+                return
```

Comment:
I could write it like this
```python
while os.path.getmtime(path) == res:
    pass
```
but I think the current way is easier to understand.